src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,155 @@
+                before.next.prev = adding;
+                before.next = adding;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < NON_NEGATIVE_INDEX || index >= size) {
+            throw new IndexOutOfBoundsException(INDEX + index + SIZE + size);
+        }
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < NON_NEGATIVE_INDEX || index >= size) {
+            throw new IndexOutOfBoundsException(INDEX + index + SIZE + size);
+        }
+        Node<T> item = getNode(index);
+        T oldDate = item.value;
+        item.value = value;
+        return oldDate;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < NON_NEGATIVE_INDEX || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> current = getNode(index);
+        unlink(current);
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (Objects.equals(object, currentNode.value)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == NON_NEGATIVE_INDEX;
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < NON_NEGATIVE_INDEX || index > size) {
+            throw new IndexOutOfBoundsException(INDEX + index + SIZE + size);
+        }
+        Node<T> current;
+        if (index < size / DIVORCE_NUMBER) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - ONE; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private void unlink(Node<T> unlinkedNode) {
+        if (unlinkedNode == head) {
+            head = unlinkedNode.next;
+        } else {
+            unlinkedNode.prev.next = unlinkedNode.next;
+        }
+        if (unlinkedNode == tail) {
+            tail = unlinkedNode.prev;
+        } else {
+            unlinkedNode.next.prev = unlinkedNode.prev;
+        }
+        size--;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(T value) {
+            this.value = value;
+        }
```

Comment:
At this inner class it would be better to make constructor for all three variables, not only value